### PR TITLE
Add a new flavor dimension "backend"

### PR DESCRIPTION
**Pull request by duswnchl**
Currently the backend for a browser engine can be selected by  userProperties.backend. This PR adds a new flavor dimension "backend" and we can choose the backend from the build variant.

So far, "chromium"/"webkit" is available only when userProperties.chromium_aar/userProperties.webkit_aar are set.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -89,15 +89,19 @@ def getMKApiKey = { ->
     return ""
 }
 
-def getBackend = { ->
-    if (gradle.hasProperty("userProperties.backend")) {
-        def engine = gradle."userProperties.backend"
-        assert engine == "gecko" || engine == "chromium" || engine == "webkit"
-        return engine
+def isChromiumAvaiable = {
```
Comment (svillar) [inline app/build.gradle:92]: Typo: Avaiable->Available

File: app/build.gradle [hunk 2]
```diff
@@ -339,23 +357,26 @@ android {
         // Create variants for China only for HVR platforms
         if (country == 'cn' && !platform.startsWith('hvr'))
             variant.setIgnore(true);
-    }
 
+        // Create variants for chromium/webkit, only when they are avialble
```
Comment (svillar) [inline app/build.gradle:361]: Typo: avialble->available

File: app/build.gradle [hunk 3]
```diff
@@ -89,15 +89,19 @@ def getMKApiKey = { ->
...
-        return engine
+def isChromiumAvaiable = {
+    if (gradle.hasProperty("userProperties.chromium_aar")) {
+        return true
     }
-    return "gecko"
+    return false
 }
 
+def isWebKitAvaiable = {
```
Comment (svillar) [inline app/build.gradle:99]: Typo: same as before

**Review by svillar — CHANGES_REQUESTED** [review]
Looking great. I'll have to test it but in the meantime I have just a few nits (typos mostly)

File: app/build.gradle (hunk 1 above)
Comment (duswnchl) [reply app/build.gradle]: Oops! Thanks for the catch :) Done.

File: app/build.gradle (hunk 2 above)
Comment (duswnchl) [reply app/build.gradle]: Done.

File: app/build.gradle (hunk 3 above)
Comment (duswnchl) [reply app/build.gradle]: Done.

File: app/build.gradle [hunk 4]
```diff
@@ -587,23 +626,21 @@ dependencies {
...
-} else {
-    dependencies {
-        // "nightly" or "beta"
-        def branch = "nightly"
-        arm64Implementation deps.gecko_view."${branch}_arm64"
-        x86_64Implementation deps.gecko_view."${branch}_x86_64"
+
+    // webkit
+    if (isWebKitAvailable()) {
+        webkitImplementation fileTree(dir: gradle."userProperties.webkit_aar", include: ['*.aar'])
```
Comment (zakharvoit) [inline app/build.gradle:643]: I think it should be "implementation"

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -12,7 +12,7 @@ jobs:
 
     strategy:
       matrix:
-        flavour: [assembleOculusvrArm64World, assembleHvrArm64World, assemblePicoxrArm64World]
+        flavour: [assembleOculusvrArm64WorldGecko, assembleHvrArm64WorldGecko, assemblePicoxrArm64WorldGecko]
```
Comment (jdapena) [inline .github/workflows/ci.yml:15]: Nit: maybe time to make this array multiline, so it is a bit more version control friendly.

**Review by jdapena — CHANGES_REQUESTED** [review]
It mostly looks good to me (apart from a minor nitpick). I'd likely prefer some refactor in build.gradle to avoid repeating same code in each backend, but that definitely needs some thinking before doing.

File: app/build.gradle [hunk 6]
```diff
@@ -477,6 +498,24 @@ android {
...
+            java.srcDirs = [
+                    'src/common/chromium'
+            ]
+        }
+
+        webkit {
+            java.srcDirs = [
+                    'src/common/webkit'
+            ]
+        }
```
Comment (svillar) [inline app/build.gradle:518]: Why is this preferable over having `'src/common/' + getBackend()` above?

File: app/build.gradle [hunk 7]
```diff
@@ -587,23 +626,28 @@ dependencies {
...
+    // gecko
+    def branch = "nightly" // "nightly" or "beta"
+    geckoImplementation deps.gecko_view."${branch}_x86_64"
+    geckoImplementation deps.gecko_view."${branch}_arm64"
+    configurations.all {
+         resolutionStrategy.capabilitiesResolution.withCapability('org.mozilla.geckoview:geckoview') {
+            def abi = getName().toLowerCase().contains('x86_64') ? 'x86_64' : 'arm64'
+            def candidate = "geckoview-${branch}-${abi}"
+            select(candidates.find { it.id.module.contains(candidate) })
+        }
```
Comment (svillar) [inline app/build.gradle:639]: Not really sure, what this is about? Could you explain why it's needed?

**Review by svillar — COMMENTED** [review]
I've a couple of doubts, once resolved it should be ready to merge.

**Review by jdapena — APPROVED** [review]

File: app/build.gradle (hunk 7 above)
Comment (svillar) [reply app/build.gradle:639]: fair enough, thanks for the explanations

**Review by svillar — APPROVED** [review]
Awesome work

**Pull request merged**